### PR TITLE
Fix #3113 Use lock in Container.Remove

### LINKED ISSUE #3113: Container.Remove() from multiple goroutines causes panic
### Describe the bug:
Calling Container.Remove() from multiple goroutines at the exact same time causes a panic
Container.Add() seems to be working fine (unless you're trying to add 10k+ items on a 2 fps display then it crashes)
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x58 pc=0x7bd2b9]

goroutine 78 [running]:
fyne.io/fyne/v2/layout.(*boxLayout).Layout(0xc0003b19d8, {0xc0004e8000, 0x7d, 0xc000167250?}, {0x48732e?, 0x0?})
        /home/ryan/go/pkg/mod/fyne.io/fyne/v2@v2.2.2/layout/boxlayout.go:63 +0x399
fyne.io/fyne/v2.(*Container).layout(...)
        /home/ryan/go/pkg/mod/fyne.io/fyne/v2@v2.2.2/container.go:186
fyne.io/fyne/v2.(*Container).Remove(0xc00037fe40, {0xb0fe98?, 0xc0003f9e00})
        /home/ryan/go/pkg/mod/fyne.io/fyne/v2@v2.2.2/container.go:139 +0x1fb
main.removeLater({0xb0fe98, 0xc0003f9e00})
        /home/ryan/Downloads/FyneContainerCrash/main.go:43 +0x54
created by main.main
        /home/ryan/Downloads/FyneContainerCrash/main.go:29 +0xf0
exit status 2
```

### To Reproduce:
Example code here: https://go.dev/play/p/pSAsuB7XU-5

Steps to reproduce the behaviour:
1. Create a container
2. Add a ton of objects to that container
3. Launch a goroutine for each object
4. Have each goroutine remove an object from the container at the same time
5. Panic

### Device (please complete the following information):
 - **OS:** Linux (amd64)
 - **Version:** 5.18.7-arch1-1
 - **Go version:** 1.18.3
 - **Fyne version:** 2.2.1

### PR DESCRIPTION
**Pull request by cedy**
### Description:

Use lock around objects slice to prevent race condition.

Fixes #3113

### Checklist:

- [x] Tests included.
- [x] Lint and formatter run with no errors.
- [x] Tests all pass.

### CODE REVIEW

File: container.go [hunk 1]
```diff
@@ -10,8 +14,9 @@ type Container struct {
 	position Position // The current position of the Container
 	Hidden   bool     // Is this Container hidden
 
-	Layout  Layout         // The Layout algorithm for arranging child CanvasObjects
-	Objects []CanvasObject // The set of CanvasObjects this container holds
+	Layout      Layout // The Layout algorithm for arranging child CanvasObjects
+	objectsLock *sync.Mutex
```
Comment (Jacalz) [inline container.go:18]: Isn’t it possible to not use a pointer here and thus avoid the need to pass a new value each time the struct is created? The current way would crash anyone using containers without the constructors.

File: container.go [hunk 2]
```diff
@@ -1,5 +1,9 @@
 package fyne
 
+import (
+	"sync"
+)
```
Comment (Jacalz) [inline container.go:5]: Let’s get this on one row please.

**Review by Jacalz — CHANGES_REQUESTED** [review]
Thanks for the fix. I have a few suggestions.

File: container.go (hunk 1 above)
Comment (Jacalz) [reply container.go:18]:
```suggestion
	objectsLock sync.Mutex
```

Comment (cedy) [reply container.go]: Definitely, no need for a pointer. Thanks

File: container.go (hunk 2 above)
Comment (cedy) [reply container.go]: 👍

**Comment by andydotxyz** [comment]
Would it be worth adding to `Add` at the same time? It will suffer a version of this crash as well I think...

**Comment by ErrorNoInternet** [comment]
> Would it be worth adding to `Add` at the same time? It will suffer a version of this crash as well I think...

Just did some testing, and it seems like the more widgets you try to add, the higher the chance of crashing. Sometimes it crashes when I'm adding more than 512 widgets (I would say a 30% chance of crashing), but when it doesn't crash, not all of the 512 widgets are added. I can only see about 10-20 widgets in the container.

So yes, adding a mutex to `Add` would be nice.

**Comment by cedy** [comment]
> Would it be worth adding to `Add` at the same time? It will suffer a version of this crash as well I think...

Indeed, added there as well

File: container.go [hunk 3]
```diff
@@ -10,8 +12,9 @@ type Container struct {
 	position Position // The current position of the Container
 	Hidden   bool     // Is this Container hidden
 
-	Layout  Layout         // The Layout algorithm for arranging child CanvasObjects
-	Objects []CanvasObject // The set of CanvasObjects this container holds
+	Layout      Layout // The Layout algorithm for arranging child CanvasObjects
+	objectsLock sync.Mutex
```
Comment (Jacalz) [inline container.go:16]: Should we just call it just `lock`?

**Review by Jacalz — CHANGES_REQUESTED** [review]
Looking good. Thanks. Just one small wording change that I think would be beneficial.

File: container.go (hunk 3 above)
Comment (cedy) [reply container.go]: 👍

**Review by Jacalz — DISMISSED** [review]
Thanks. This looks sensible to me. I just want a second opinion before we merge as I'm only a novice when it comes to mutexes.

**Comment by Jacalz** [comment]
Ah, looks like you have some formatting that needs to be fixed as well.

**Comment by cedy** [comment]
> Ah, looks like you have some formatting that needs to be fixed as well.

yeah, annoying when go fmt should run on file save but it doesn't, fixed now :)

**Comment by Jacalz** [comment]
It looks like tests are failing as well.

**Comment by cedy** [comment]
> It looks like tests are failing as well.

Seem to pass now after rebase

**Review by Bluebugs — APPROVED** [review]
This does improve things significantly. A good step forward.

**Review by Jacalz — APPROVED** [review]
Looks like this might have been waiting for my review. Sorry for the wait.

**Pull request merged**